### ISSUE TITLE
depfiles not working for incbin asm statements

### ISSUE THREAD

**Issue by kusma** [mesonbuild/meson#9849]
**Describe the bug**

The dependency-tracking method Meson's Ninja backend uses doesn't keep track of the `.incbin <source>` assembly directive. The same issue probably also exists for the `.include <source>` directive, but that can be worked around using the C pre-processor on top.

Arguably the problem lies in GCC, but Meson doesn't seem to provide any escape-hatch that can be used to add back the missing dependencies.

**To Reproduce**

Create the following project:

`foo.S`:
```asm
.section .data

.global foo_bin
foo_bin:
.incbin "foo.bin"
```
`foo.bin`:
```
abcdef
```
`main.c`:
```c
#include <stdio.h>
extern char foo_bin[];
int main()
{
       printf("foo_bin: \"%.*s\"\n", 4, foo_bin);
       return 0;
}
```
`meson.build`:
```
project('meson-incbin-test', 'c')
executable('incbin-test', 'main.c', 'foo.S')
```

Now, run
1. `meson setup build`
2. `ninja -C build/`
3. `echo "foobar" > foo.bin`
4. `ninja -C build/`
5. `build/incbin-test`
6.  Observe how `foo_bin: "abcd"` is printed.

**Expected behavior**
The output `foo_bin: "foob"` should be printed on screen 

**system parameters**

Happens both for cross and native builds.
Ubuntu 21.10
Python 3.9.7
Meson 0.61.0
Ninja 1.10.1

**Comment by eli-schwartz** [comment]
Add `foo.bin` as a source file, meson considers it an unknown filetype and treats it the same way it treats `*.h` files -- by making it an order-only dependency.

**Comment by kusma** [comment]
@eli-schwartz:
That gives the following error:
```
ERROR: No specified compiler can handle file foo.bin
```

Can you please reopen?

**Comment by eli-schwartz** [comment]
Actually I'm not sure whether that will recompile the asm file itself... Hmm.

**Comment by eli-schwartz** [comment]
Huh, i know this works for stuff like .inc so I'm not sure why it fails for .bin

**Comment by kusma** [comment]
Dunno, doesn't seem to work for .inc for me either:

> ERROR: No specified compiler can handle file foo.inc

**Comment by kusma** [comment]
Seems [CMake has had similar issues](https://gitlab.kitware.com/cmake/cmake/-/issues/20426), which I guess explains things a bit:

Here's what I generate if I run the compile-step from the example:

```
incbin-test.p/foo.S.o: ../foo.S /usr/include/stdc-predef.h
```

That's using the C front-end.

However, if I use the GNU AS frontend, like so: `as foo.S -MD foo.S.o.d -o foo.S.o`, then I get the expected (ignoring folders and whatnot, because I'm just YOLOing things directly in the source folder here):
```
foo.S.o: foo.bin foo.S
```

**Comment by kusma** [comment]
I don't think unconditionally using AS as the frontend instead of CC is going to do the right thing for .S sources, because those are supposed to be pre-processed. But I guess we should use it for .s sources (just checked, renaming `foo.S` to `foo.s` and updating the name in `build.meson` didn't change anything; seems like we currently treat .s and .S the same way... probably not a good idea?).

**Comment by kusma** [comment]
I suppose a solution could be to use AS as the front-end and pre-process .S files as an explicit step first?

**Comment by kusma** [comment]
Knowing that, I realize I can work around this myself, by compiling .S to .o using a `custom_target()` with depfile.

```patch
diff --git a/foo.S b/foo.S
index 31e5662..11c9074 100644
--- a/foo.S
+++ b/foo.S
@@ -2,4 +2,4 @@

 .global foo_bin
 foo_bin:
-.incbin "foo.bin"
+.incbin "../foo.bin"
diff --git a/meson.build b/meson.build
index 47401eb..f66b07f 100644
--- a/meson.build
+++ b/meson.build
@@ -1,2 +1,10 @@
 project('meson-incbin-test', 'c')
-executable('incbin-test', 'main.c', 'foo.S')
+foo_o = custom_target('foo.o',
+  input: 'foo.S', output: 'foo.S.o',
+  depfile: 'foo.S.o.d',
+  command : [
+    find_program('as'), '@INPUT@',
+    '-I', '@CURRENT_SOURCE_DIR@',
+    '-MD', '@DEPFILE@',
+    '-o', '@OUTPUT@']
+)
+executable('incbin-test', 'main.c', foo_o)
```

Still, I think this should be fixed properly in the long run.

**Comment by dcbaker** [comment]
Out of curiosity, does the same happen if you use clang?

**Comment by kusma** [comment]
> Out of curiosity, does the same happen if you use clang?

Yep, exact same behavior. However, my work-around seems to still depend on as from GNU binutils instead of llvm-as in that case, because it seems Meson doesn't support setting a custom $AS envvar (like it does to select the compiler through $CC)...

So I guess that's another argument for fixing it properly as well...

**Comment by kusma** [comment]
Also, reading the [GCC docs](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#Preprocessor-Options), I don't actually think this is technically speaking a GCC bugs. First of all, these flags are all documented under the section for "Preprocessor Options". But the docs for the -M families of switches also seems to be pretty clear that this is about the pre-processor only...

However, the vaguer wording in the [AS docs](https://sourceware.org/binutils/docs/as/MD.html#MD) seems fairly clear that this does what we want.

So yeah, I think that's a pretty clear smoking gun here, pointing at this being indeed a Meson bug.

**Comment by dcbaker** [comment]
It looks like then we need to call gas directly instead of through gcc/clang?

**Comment by kusma** [comment]
> It looks like then we need to call gas directly instead of through gcc/clang?

Yeah, exactly. But .S sources (uppercase) should be preprocessed first, as opposed to .s sources (lowercase)...

**Comment by dcbaker** [comment]
Sigh, a thing we try *so* hard not to support (because of course, all OSes but Linux, and Linux in the future it seems like)

**Comment by kusma** [comment]
@dcbaker: I'm not sure I understand what you're referring to... We already support both .s and .S, we just use the wrong tool for these (the compiler instead of the preprocessor and assembler). What is it that we're trying hard not to support?

**Comment by dcbaker** [comment]
We work really hard to *not* consider capitalization of file extensions, since that doesn't work on MacOS or windows, or Linux with the new case-insensitive file system options that seem to be on the horizon as the new norm.

**Comment by kusma** [comment]
Aha, I understand. I guess in this case we could use the capitalization from the meson.build file rather than the filesystem, though?

**Comment by xclaesse** [comment]
> We work really hard to _not_ consider capitalization of file extensions, since that doesn't work on MacOS or windows, or Linux with the new case-insensitive file system options that seem to be on the horizon as the new norm.

Case insensitive does not mean files does not have case. You can still check if the file has capital S or not even on Windows.

**Comment by kusma** [comment]
> Case insensitive does not mean files does not have case. You can still check if the file has capital S or not even on Windows.

One danger in respecting the on-disk casing on Windows is that a file might end up with a different casing than desired, for instance if it was first checked in with the wrong case, and then corrected. Often such a move will fail on Windows, because the destination file already exists, but some version control software ignores those failures. So, it might work most of the time, but it doesn't strike me as robust. But I think checking the casing used in the build-file should be.

> That said, is there any downside of just going through the preprocessor all the time?

Yes, there certainly is. Lines starting with # is treated as comments for i386 and x86_64 architectures. That means that any comment will be interpreted as a pre-processor statements.

**Comment by kusma** [comment]
> One danger in respecting the on-disk casing on Windows [...]

Just to be clear, this doesn't mean that I'm sure that relying on the file-system casing is the the wrong thing to do. Maybe this is a good way to let the user know that something's up and they need to fix their files. Just thought I'd mention it.

**Comment by xclaesse** [comment]
> But I think checking the casing used in the build-file should be.

That's what matters indeed, so there is no issue at all to rely on case. Isn't gcc already doing that anyway when meson gives a .S or .s file?

**Comment by kusma** [comment]
Good question, I should check what GCC does on a case-insensitive filesystem when given the incorrect casing on the command-line... My *guess* would be that it uses the casing from the command-line, because checking the casing on disk kinda requires bending over backwards a bit...

**Comment by kusma** [comment]
Just gave it a quick test, and yeah, GCC uses the casing from the command-line:
```
$ echo '#include "no-file.h"' > test.s
$ gcc -c test.s
$ gcc -c test.S
test.S:1:10: fatal error: no-file.h: No such file or directory
    1 | #include "no-file.h"
      |          ^~~~~~~~~~~
compilation terminated.
```

**Comment by xclaesse** [comment]
I did a bit of investigation here, what I found is we need to pass `-Wa,-MD,foo.d` compiler argument to gcc. That will pass `-MD foo.d` down to the assembler to generate assembler level dependencies. Problem is that would generate 2 depfile for a single ninja target and I don't think ninja supports that.

The good news is we don't have to deal with GAS directly, we can keep using gcc frontend which simplifies a lot the problem because it means we can keep using CFAGS/c_args. All we need to do is invoke gcc twice: First to preprocess .S to .s and that will generate preprocessor level depfile, then a 2nd time to compile .s to .o with the extra argument `-Wa,-MD,foo.d` and that will generate the assembly level depfile.

Bonus side effect of supporting that is on Windows we can reuse the same preprocessing code path to preprocess files with cl.exe before compiling with ml64.exe as requested in https://github.com/mesonbuild/meson/issues/9889.

**Comment by xclaesse** [comment]
Aaargh, it's not working with clang:
```
clang: error: unsupported argument '-MD' to option 'Wa,'
clang: error: unsupported argument 'cpp-asm.exe.p\meson-generated_retval-x86_64.S.s.obj.d' to option 'Wa,'
```

**Comment by xclaesse** [comment]
Ok, there is actually no way to support this with clang at all, llvm-as does not support -MD option.

Reported the bug in llvm: https://bugs.chromium.org/p/llvm/issues/detail?id=25. In the meantime, my PR https://github.com/mesonbuild/meson/pull/10149 fix it for GCC only.
